cfme/infrastructure/host.py

```diff
@@ -92,15 +92,29 @@ def __init__(self, **kwargs):
             super(Host.Credential, self).__init__(**kwargs)
             self.ipmi = kwargs.get('ipmi')
 
-    def update(self, updates, validate_credentials=False):
+    def update(self, updates, validate_credentials=False, action='edit_from_details'):
         """Updates a host in the UI. Better to use utils.update.update context manager than call
         this directly.
 
         Args:
            updates (dict): fields that are changing.
+           action (str): denotes additional functionality. Expecting edit_from_details,
```

Comment:
I see how you're trying to consolidate on the test function calls by pushing options into `action` here, but its inconsistent with the rest of the entity update methods, and I don't think its something we want to build into the framework in this way.

Just using bare strings is going to be an issue, especially if we try to apply this kwarg to other entity update methods.

That aside, nearly all entity update methods are written to navigate to the entity `Edit` destination (which uses `Details` as prerequisite), and then fill the form.  Test cases parametrize on the nav path, and navigate + form fill directly if not navigating from details.

I don't think that the `Updateable` mixin / `update` contextmanager will actually handle the `cancel` action here either - unless an exception is raised on `cancel=True`, the contextmanager will return an updated entity instance. The entity attributes will no longer match what's in MIQ.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/update.py#L54

Reply:
I was pretty confident the cancel stuff was working. Maybe I missed something, but I tested this when I don't click save in update, the custom_ID is the same as it was before I ran the test. Anyway, the bigger issue is how do we want to implement all of the cases without actions? Do I add a bunch or options kwargs instead? Suggestions? Maybe we should focus back on PR#9761 and work on this after it's merged? Thoughts?

Reply:
I am going to implement actions as a few kwargs. If there is another suggested solution please comment.

Reply:
I had originally written this as it's own test. This is a customer BZ. I am beginning to have many one of tests for customer BZs and was trying to consolidate into another test that made sense logically. Can I get a little more on why exactly update is the wrong place for this? I understand I my initial implementation with strings is not ideal (I have already updated to use kwargs) but why not in update? If I put this functionality somewhere else, it's going to be a lot of duplication of host.update.

Reply:
I'd have to agree with @tpapaioa here, I don't think that the update method is the appropriate place for these more complex interaction paths.

It creates an inconsistent API for handling update on the entity, compared to all the other Entity update methods.  These methods are generally just implemented to handle the standard positive path actions for editing an entity through the UI - where `cancel` kwarg is just controlling whether `Save` or `Cancel` is clicked after filling the form.  Many of these entity methods also include a `from_details` kwarg, that controls how the automation gets to the edit page.  That should not be changed here to `from_hosts` as again that makes for an inconsistent API for the entity CRUD methods.

I don't think this is a pattern of behavior that we should push to all the other Entity classes with update methods, so I think its better to capture this specific behavior within the test as Tasos suggested.

There will be some duplication of what's in host.update within the test, but not that much, and it will be specific to the UI behavior that we're trying to explicitly test.  The specific navigation cases that you're trying to test don't require you to duplicate all of the behavior in update, just what's relevant to the test. We're not trying to capture a large permutation matrix of the actions, just a few specific paths.  This may make the test code a bit longer, but the behavior is specific to this test coverage - not behavior that we'll want to regularly replicate on a host for other test coverage.

To summarize, please:
1. Keep `from_details` kwarg on `Host.update` method
2. Add `cancel` kwarg to `Host.update`, given that it works properly in the context manager
3. Update the test to take the navigation steps for the actions that you want to test

In updating the test - the action paths don't have to be parametrized, as several can be covered in one test case flow. You can update with cancel, update from details, update from hosts, update and nav away with no changes, update and nav away with changes, and then delete, all in one flow.

Reply:
Updated as suggested.